fixes #4386 - implement gem friendly_id to simply find by id, name, label, etc

isratrade opened the pull request.

isratrade commented:
rebased

isratrade commented:
@domcleal, is the failure here due to RPM builds?

domcleal commented:
@isratrade not entirely, there are also failures in a core functional test: http://ci.theforeman.org/job/test_develop_pr_core/472/.  Ignore the RPM failure for now, we'll need to update the spec file and build friendly_id if we haven't already.

isratrade commented:
@domcleal, failing text on hosts controller is fixed.  Re-committed.

isratrade commented:
rebased and all tests pass except rpm packaging build

isratrade commented:
ping to any reviewers

isratrade commented:
@ohadlevy, can you review?

isratrade commented:
@ohadlevy, @domcleal, what are feelings of merging this before the feature freeze?

domcleal commented:
Frankly, I don't know what to do here as I'm not familiar with friendly_id or the pros and cons of each way of writing the API, sorry.

ohadlevy commented:
@isratrade for the first question, I dont have any feelings about it, it might change depnding on which feature freeze you mean :8ball: 

sounds like a deep dive?

isratrade commented:
@ohadlevy, @domcleal, I can do a deep dive on the friendly_id gem. 

In short, it allows us to do this

``` ruby
Host.find('f18.sat.lab.tlv.redhat.com') 
```

rather than this

``` ruby
Host.find_by_name('f18.sat.lab.tlv.redhat.com') 
```

Currently, we manually check if the params[:id] or params[:nested_id] is an integer or string and we also have different `find_by_` logic in UI `application_controller` and API `base_controller` 

for UI

``` ruby
  def find_by_name
    not_found and return if params[:id].blank?

    name = controller_name.singularize
    cond = "find" + (params[:id] =~ /\A\d+(-.+)?\Z/ ? "" : "_by_name")
    not_found and return unless instance_variable_set("@#{name}", resource_base.send(cond, params[:id]))
  end
```

for API

``` ruby

    def find_resource(controller = controller_name)
      resource = resource_identifying_attributes.find do |key|
        next if key=='name' and (params[:id] =~ /\A\d+\z/)
        method = "find_by_#{key}"
        id = key=='id' ? params[:id].to_i : params[:id]
        scope = resource_scope(controller)
        if scope.respond_to?(method)
          (resource = scope.send method, id) and break resource
        end
      end
```

This PR in particular simplifies the API find logic for nested routes. API controllers have method `resource_identifying_attributes`
app/controllers/api/v2/users_controller.rb

``` ruby
      def resource_identifying_attributes
        %w(login id)
      end
```

api/v2/operatingsystems_controller.rb

``` ruby
      def resource_identifying_attributes
        %w(to_label id)
      end
```

Several times for a nested route,  I would get the error "no method find_by_login" or "no method find_by_to_label"

I also want to use the gem with "scoped" features to simplify the find_by_name when name isn't unique such as with smart_class_parameters.  I think it would also be helpful for Katello (although admittedly haven't looked into it) when find_by_name isn't unique but is unique within organization.

dLobatog commented:
Generally I like how friendly_id would simplify the code, this needs a rebase though.

isratrade commented:
@eLobato, thanks for reviewing this. I just rebased. I also reviewed it with @ehelms as he has an open PR #1596 so allow plugins to use `find_resource`

isratrade commented:
[test] again due to random failure YAML output for rundeck

boaz0 commented:
:+1:

dLobatog commented on app/controllers/concerns/foreman/controller/taxonomies_controller.rb:
```diff
@@ -181,13 +181,16 @@ def taxonomy_class
       when 'locations'
         @taxonomy = @location = Location.find(params[:id])
     end
+    not_found and return false unless @taxonomy
```
If this necessary? If @taxonomy isn't found the controller could handle the not_found, just like other controllers using the regular FindCommon code don't need this line.

isratrade replied:
@eLobato, I agree. Line is now removed.

dLobatog commented:
:+1: for the review, please rebase so this can be tested

isratrade commented:
@eLobato, can you review this again?

dLobatog commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -21,9 +21,11 @@ def create
   end
 
   def edit
+    @proxy = @smart_proxy
```
Why not using `@smart_proxy` in the views directly?

dLobatog commented:
I think it looks good, it needs a rebase and I have to do some more testing, probably ready to merge after rebase.

isratrade commented:
@eLobato, re-based.

dLobatog commented:
It needs rebasing again, sorry... some issues I've found:
- No friendly_id for external_usergroups.
- Some commit about another issue slipped in
- There are other controllers could use find instead of find_by_name in several places, grep for "find_by_name" in these,
  - api/v2
    - lookup_keys_common, reports, operatingsystems
  - concerns
    - api/import_puppetclasses_common, api/v2/lookup_keys_common

Commit:
fixes #4386 - implement gem friendly_id to simply find by id, name, label, etc

isratrade commented:
@eLobato, 
- added friendly_id for external_usergroups
- changed some `find_by_name` to `find`
- did not change the instances of `find_by_name` where I want value to return `nil` and not throw exception as `.find()` would do. I was hoping there would be an option in `friendly_id` to fail gracefully if not found, but doesn't seem to exist (at least not in friendly_id 4).

isratrade commented:
@eLobato, another options to do something like this with `rescue nil`

``` ruby
  def find_optional_environment
-    @environment   = Environment.authorized(:view_environments).find_by_id(@env_id.to_i) if @env_id.to_i > 0
-
-    @environment ||= Environment.authorized(:view_environments).find_by_name(@env_id)
-    @environment

+    @environment = Environment.authorized(:view_environments).find(@env_id) rescue nil
  end
```

isratrade commented:
[test] only failure is 'YAML output for rundeck'

dLobatog commented on app/controllers/concerns/api/import_puppetclasses_common_controller.rb:
```diff
@@ -108,9 +107,7 @@ def get_environment_id
-    @environment   = Environment.authorized(:view_environments).find_by_id(@env_id.to_i) if @env_id.to_i > 0
-    @environment ||= Environment.authorized(:view_environments).find_by_name(@env_id)
-    @environment
+    @environment = Environment.authorized(:view_environments).find(@env_id) rescue nil
```
I think you only want to `rescue ActiveRecord::RecordNotFound:`?

isratrade commented:
@eLobato, I made more changes in a separate commit including removing `resource_identifying_attributes` from `find_resource` in the API and adding `title` to the operating systems table, so we don't need to use `self.find_by_to_label`
https://github.com/theforeman/foreman/blob/develop/app/models/operatingsystem.rb#L148-L157

Commit:
squash this commit

dLobatog commented:
Squashed and merged as https://github.com/theforeman/foreman/commit/8b737c9c7648b3726dadb3b2e4708fcb43af02a8 , thank you @isratrade !
